### PR TITLE
Hotline Bling

### LINKED ISSUE #273: Deprecate use of UniqueTagSet class
> can't we already remove it? Since its not being used anywhere in the whole codebase

_Originally posted by @qwoprocks in https://github.com/AY2021S1-CS2103T-F11-4/tp/pull/268#discussion_r514080138_

### LINKED ISSUE #293: [PE-D] Phone numbers should allow spaces according to UG but fails validity check.
When adding a client with a phone number with spaces, it is recognised as an invalid number. However, according to the UG, I should be able to add phone numbers with spaces.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: khoongwk/ped#1

### LINKED ISSUE #300: [PE-D] Invalid phone numbers
Phone numbers should have a maximum limit of 15 digits.

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: rolandyuwy/ped#3

### PR DESCRIPTION
**Pull request by rtshkmr**
## Description

Changed the regex to `(^[\+][0-9]{1,3}[-\s\.]?[0-9]{2,12}$)|(^[0-9]{3,12}$)`

Fixes #300 
Fixes #293 
Fixes #273 

Documentation changes have also been made: 
* added an appendix to explain the Country codes specification and modified some links (added a link to the official country code finder tool)

## Testing
https://www.regextester.com/
international standard: https://en.wikipedia.org/wiki/E.164

valid: 
123456
+65 81867752
+6581867752
+6581867752
+65 81867752
+123 81867752
+12381867752
+12381867752
+123 81867752
81867752
123456789123456
+6 3456

accepted but weird: 
+65-81867752
+65 3456
+653456

expected failure: 
99
65 81867742
1234567891234567
+65123456789123456
+1234 81867752
0
+65) 81867752
(+65) 81867752
+(65)81867752

## Remarks

really needs to be well tested for fear of undiscovered bugs :fearful:

### CODE REVIEW

File: src/test/java/seedu/address/model/client/PhoneTest.java [hunk 1]
```diff
@@ -23,21 +23,49 @@ public void constructor_invalidPhone_throwsIllegalArgumentException() {
...
+        assertTrue(Phone.isValidPhone("123456789123456")); // max length of 15 numbers in total (country code+num)
+
+        // weird but valid numbers:
+        assertTrue(Phone.isValidPhone("+(65)-81867752"));
+        assertTrue(Phone.isValidPhone("+65 3456"));
+        assertTrue(Phone.isValidPhone("+653456"));
+
+
+
+
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/PhoneTest.java]: extra newlines

File: src/main/java/seedu/address/model/client/Phone.java [hunk 2]
```diff
@@ -11,8 +11,9 @@
 
 
     public static final String MESSAGE_CONSTRAINTS =
-            "Phone numbers should only contain numbers, and it should be at least 3 digits long";
-    public static final String VALIDATION_REGEX = "\\d{3,}";
+            "Phone numbers should have at least 3 digits and at most 15 digits, please refer to the UserGuide for"
```
Comment (qwoprocks) [inline src/main/java/seedu/address/model/client/Phone.java]:
```suggestion
            "Phone numbers should have at least 3 digits and at most 15 digits, please refer to our UserGuide for"
```

File: docs/UserGuide.md [hunk 3]
```diff
@@ -599,4 +604,23 @@ This is what to expect when you type in the `help` command or press `F1`:
...
+[Here's the Official Search tool](https://www.iso.org/obp/ui/#search) for country codes.
+
+### Phone Numbers
+
+**_TBM_**'s format for phone numbers is **_based_** on the [E.164 standard](https://en.wikipedia.org/wiki/E.164) as recommended by the _ITU Telecommunication Standardization Sector (ITU-T)_.
+**_However_**, **_TBM_** doesn't allow for spaces in between phone number fields. For example, `+65 8186 7742` may follow this international, but **_TBM_ requires you to remove spacing within the phone number field** so the correct way to input it would be `+65 81867742`. 
+
+<p align="center">
+   <strong><i>{End of Appendix B, please return to the <a href="#table-of-contents">Table of Contents</a> } </i
+   ></strong>
```
Comment (qwoprocks) [inline docs/UserGuide.md]: \<i\> tag broken

File: docs/UserGuide.md [hunk 4]
```diff
@@ -599,4 +604,23 @@ This is what to expect when you type in the `help` command or press `F1`:
    <strong><i>{End of Appendix A, please return to the <a href="#table-of-contents">Table of Contents</a> } </i
```
Comment (qwoprocks) [inline docs/UserGuide.md]: \<i\> tag broken

File: docs/UserGuide.md [hunk 5]
```diff
@@ -530,10 +535,10 @@ Action | Format, Examples
...
 **NAME** | `n/` | Names should only contain alphanumeric characters and spaces, and it should not be blank. <br> e.g. `n/John Doe`
-**PHONE_NUMBER** | `p/` | Phone numbers should only contain numbers, and it should be at least 3 digits long. <br> e.g. `p/9123 4567`
+**PHONE_NUMBER** | `p/` | Phone numbers should have at least 3 digits and at most 15 digits. <br> Please refer to [Appendix B](#appendix-b-international-standards-followed) for a comprehensive description. <br> e.g. `p/81867752 or +65 81867752` 
 **EMAIL** | `e/` | Emails should be of the format local-part@domain. <br> e.g. `e/katya@yahoo.com`
 **ADDRESS** |`a/` | Addresses can take any values, and it should not be blank. <br> e.g. `Vladivostok, Nevelskogo, bld. 15, appt. 256`
-**COUNTRY_CODE** | `c/` | A 2-letter country code that follows the ISO3166 specification. [List](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) of country codes. <br> e.g. `c/SG` (Singapore)
+**COUNTRY_CODE** | `c/` | A 2-letter country code that follows the ISO3166 specification  <br> [Here's the official search tool](https://www.iso.org/obp/ui/#search) for country codes. <br> e.g. `c/SG` (Singapore)
```
Comment (qwoprocks) [inline docs/UserGuide.md]: extra space before br

File: docs/UserGuide.md [hunk 6]
```diff
@@ -530,10 +535,10 @@ Action | Format, Examples
 Parameter | Prefix | Constraints, Examples
 --------|------------------| ----------
 **NAME** | `n/` | Names should only contain alphanumeric characters and spaces, and it should not be blank. <br> e.g. `n/John Doe`
-**PHONE_NUMBER** | `p/` | Phone numbers should only contain numbers, and it should be at least 3 digits long. <br> e.g. `p/9123 4567`
+**PHONE_NUMBER** | `p/` | Phone numbers should have at least 3 digits and at most 15 digits. <br> Please refer to [Appendix B](#appendix-b-international-standards-followed) for a comprehensive description. <br> e.g. `p/81867752 or +65 81867752`
```
Comment (qwoprocks) [inline docs/UserGuide.md]:
```suggestion
**PHONE_NUMBER** | `p/` | Phone numbers should have at least 3 digits and at most 15 digits. <br> Please refer to [Appendix B](#appendix-b-international-standards-followed) for a comprehensive description. <br> e.g. `p/81867752` or `p/+65 81867752` 
```

File: docs/UserGuide.md [hunk 7]
```diff
@@ -219,8 +222,10 @@ Example:
 
 ### Saving data
 
+
```
Comment (qwoprocks) [inline docs/UserGuide.md]: extra newline

File: docs/UserGuide.md [hunk 8]
```diff
@@ -111,6 +111,9 @@ Adds a new client to **_TBM_**.
 
 Format: `client add n/NAME p/PHONE_NUMBER e/EMAIL a/ADDRESS c/COUNTRY_CODE tz/TIMEZONE [ce/CONTRACT_EXPIRY_DATE]`
 
+Please refer to [Appendix B](#appendix-b-international-standards-followed) to understand the International Standards that **_TBM_** follows for phone numbers and country codes.
+
```
Comment (qwoprocks) [inline docs/UserGuide.md]: extra newline

File: docs/UserGuide.md [hunk 9]
```diff
@@ -599,4 +604,23 @@ This is what to expect when you type in the `help` command or press `F1`:
...
+### Country Codes
+
+**_TBM_** follows 2-letter country codes that follow the ISO3166 specification.
+[Here's the Official Search tool](https://www.iso.org/obp/ui/#search) for country codes.
+
+### Phone Numbers
+
+**_TBM_**'s format for phone numbers is **_based_** on the [E.164 standard](https://en.wikipedia.org/wiki/E.164) as recommended by the _ITU Telecommunication Standardization Sector (ITU-T)_.
+**_However_**, **_TBM_** doesn't allow for spaces in between phone number fields. For example, `+65 8186 7742` may follow this international, but **_TBM_ requires you to remove spacing within the phone number field** so the correct way to input it would be `+65 81867742`.
```
Comment (qwoprocks) [inline docs/UserGuide.md]: grammatical error: "may follow this international"

**Review by qwoprocks — CHANGES_REQUESTED** [review]

File: docs/UserGuide.md (hunk 9 above)
Comment (rtshkmr) [reply docs/UserGuide.md]: woops missed out a "standard" there

File: docs/UserGuide.md [hunk 10]
```diff
@@ -596,7 +598,34 @@ This is what to expect when you type in the `help` command or press `F1`:
   <p align="center"><i>Figure 6. Help Window</i></p>
 
 <p align="center">
-   <strong><i>{End of Appendix A, please return to the <a href="#table-of-contents">Table of Contents</a> } </i
-   ></strong> 
+   <strong><i>{End of Appendix A, please return to the <a href="#table-of-contents">Table of Contents</a> }</i></strong>
```
Comment (qwoprocks) [inline docs/UserGuide.md]:
```suggestion
   <strong><i>{End of Appendix A, please return to the <a href="#table-of-contents">Table of Contents</a>}</i></strong> 
```

File: docs/UserGuide.md [hunk 11]
```diff
@@ -530,10 +532,10 @@ Action | Format, Examples
...
 **NAME** | `n/` | Names should only contain alphanumeric characters and spaces, and it should not be blank. <br> e.g. `n/John Doe`
-**PHONE_NUMBER** | `p/` | Phone numbers should only contain numbers, and it should be at least 3 digits long. <br> e.g. `p/9123 4567`
+**PHONE_NUMBER** | `p/` | Phone numbers should have at least 3 digits and at most 15 digits. <br> Please refer to [Appendix B](#appendix-b-international-standards-followed) for a comprehensive description. <br> e.g. `p/81867752` or `p/+65 81867752` 
 **EMAIL** | `e/` | Emails should be of the format local-part@domain. <br> e.g. `e/katya@yahoo.com`
 **ADDRESS** |`a/` | Addresses can take any values, and it should not be blank. <br> e.g. `Vladivostok, Nevelskogo, bld. 15, appt. 256`
-**COUNTRY_CODE** | `c/` | A 2-letter country code that follows the ISO3166 specification. [List](https://en.wikipedia.org/wiki/List_of_ISO_3166_country_codes) of country codes. <br> e.g. `c/SG` (Singapore)
+**COUNTRY_CODE** | `c/` | A 2-letter country code that follows the ISO3166 specification <br> [Here's the official search tool](https://www.iso.org/obp/ui/#search) for country codes. <br> e.g. `c/SG` (Singapore)
```
Comment (qwoprocks) [inline docs/UserGuide.md]: Think we should change this back to the wikipedia one, since the official search tool is an alternative

**Review by qwoprocks — DISMISSED** [review]
Small comments

File: docs/UserGuide.md (hunk 11 above)
Comment (rtshkmr) [reply docs/UserGuide.md]: removed all of the wikipedia links to the common This [finding tool](https://www.countrycode.org/) to find both country codes and iso codes.

File: docs/UserGuide.md [hunk 12]
```diff
@@ -438,7 +440,7 @@ Thailand
 ### Deleting notes for a country: `country note delete`
 
 Deletes a note that is associated with a specific country at the given index based on the **current view** of the [list panel displaying country notes](#ui-when-viewing-country).
-
+from this [finding tool](https://www.countrycode.org/)
```
Comment (qwoprocks) [inline docs/UserGuide.md]: random line here that is out of place

File: src/test/java/seedu/address/model/client/PhoneTest.java [hunk 13]
```diff
@@ -23,31 +23,61 @@ public void constructor_invalidPhone_throwsIllegalArgumentException() {
...
+        assertFalse(Phone.isValidPhone("+1234 81867752")); // more than 3 digits for country code
+        assertFalse(Phone.isValidPhone("+ 81867752")); // invalid country code passed in
         assertFalse(Phone.isValidPhone("phone")); // non-numeric
         assertFalse(Phone.isValidPhone("9011p041")); // alphabets within digits
         assertFalse(Phone.isValidPhone("9312 1534")); // spaces within digits
+        assertFalse(Phone.isValidPhone("+65 9312 1534")); // spaces within the phone number
+        assertFalse(Phone.isValidPhone("1234567891234567")); // exceeds 15 digits total
+        assertFalse(Phone.isValidPhone("+65123456789123456")); // exceeds 15 digits total
+        assertFalse(Phone.isValidPhone("+(65) 81867752")); // exceeds 15 digits total
+        assertFalse(Phone.isValidPhone("+0 00")); // exceeds 15 digits total
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/PhoneTest.java]: Comments not updated

File: src/test/java/seedu/address/model/client/PhoneTest.java [hunk 14]
```diff
@@ -23,31 +23,61 @@ public void constructor_invalidPhone_throwsIllegalArgumentException() {
...
         assertFalse(Phone.isValidPhone(" ")); // spaces only
         assertFalse(Phone.isValidPhone("91")); // less than 3 numbers
+        assertFalse(Phone.isValidPhone("65 81867742")); // country code not prepended with `+` char
+        assertFalse(Phone.isValidPhone("+1234 81867752")); // more than 3 digits for country code
+        assertFalse(Phone.isValidPhone("+ 81867752")); // invalid country code passed in
         assertFalse(Phone.isValidPhone("phone")); // non-numeric
         assertFalse(Phone.isValidPhone("9011p041")); // alphabets within digits
         assertFalse(Phone.isValidPhone("9312 1534")); // spaces within digits
+        assertFalse(Phone.isValidPhone("+65 9312 1534")); // spaces within the phone number
+        assertFalse(Phone.isValidPhone("1234567891234567")); // exceeds 15 digits total
```
Comment (qwoprocks) [inline src/test/java/seedu/address/model/client/PhoneTest.java]: Should this be 12 digits instead, since it is without country code

**Review by qwoprocks — APPROVED** [review]
Approved, but please fix the stuff commented

**Pull request merged**